20_2 DataGrid Custom Summaries: rewrite (#628)
DevExpress/devextreme-demos: JSDemos/Demos/DataGrid/CustomSummaries/description.md

> DataGrid can apply custom server-side and client-side functions to calculate total and group summaries. If the dataset is small, you should choose the client-side implementation.

Reviewer: > If the dataset is small, you should choose the client-side implementation.

I'm not sure saying that one *should* choose the client-side implementation in this case is correct. Client-side calculation is okay on small datasets, but so is server-side calculation.

> This demo shows how to add such a summary: it calculates the sum of sale amounts for all selected rows. To achieve this behavior, we have done the following:

Reviewer: You're describing this particular use case with selection, while it is better to describe things in general and then say how they relate to the use case shown in the demo.

> 1. Set the summary object's [summaryType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#summaryType) to *"custom"*

Reviewer: We haven't said to add an object to the totalItems or groupItems array.

> 2. Specify the custom summary's [name](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#name)

Reviewer: The name will be used to recognize the summary item within the calculateCustomSummary function. It's important.

> 3. Add the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function

Reviewer: No details are given about this function.

> 4. Implemented the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler to reload the DataGrid on each selection change.

Reviewer: This relates only to the use case in this demo, we shouldn't give it as a step.

> If the dataset is large, you should implement a server-side function instead because its performance is higher (the client does not have to download the data). For details, refer to the [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation) article.

Reviewer: I also suggest shuffling the structure a bit (don't use it as is - I haven't put much thought into wording):

DataGrid can calculate so-and-so summaries. To configure client-side calculation, do the following:

1.
1.
1.
...

In this demo, custom client-side summary calculates something.

Client-side summaries are only good for small datasets. If the dataset is large, ...

> Custom summaries should be used if you want to implement your own logic. To add them on the client side, do the following:

Reviewer: - Why "should"?
- "your own logic" of what?
- "To add them on the client side" - where else can I add them? Nothing have been said about server-side summaries yet.
- If I "add" custom summaries on the client side, does it mean they will be calculated on the client side?

> DataGrid can calculate custom summaries on the client or server side. In this demo, we have implemented the client-side logic. The instructions below explain how to do this in your code:

Reviewer 2: ```suggestion
The DataGrid can calculate custom summaries on the client or server side. In this demo, we have implemented client-side logic. The instructions below explain how to do this in your code:
```

> 1. Add a summary configuration object to the [summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/).[groupItems](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/groupItems/) or [summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/).[totalItems](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/) array

Reviewer: If a list item is a sentence, it should end with a full stop.

2. Set the summary object's [summaryType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#summaryType) to *"custom"*

3. Specify the custom summary's [name](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#name)         

> Use this value to identify the summary item inside the **calculateCustomSummary** function.

Reviewer: We haven't said anything about the `calculateCustomSummary` function yet.

1. Specify a custom aggregate logic in the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function          

> Implement the necessary calculation stages to change the resulting summary's value (the parameter's totalValue field).

Reviewer: What stages?

> In this demo, the custom summary displays the sum of sale amounts for selected rows only. To reload the DataGrid on each selection change, the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler is implemented.

Reviewer: - "the custom summary displays" is it a client-side or server-side summary?
- "To reload the DataGrid on each selection change" - why is the reload required?

> 1. Calculate the resulting value in the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function. Implement the necessary calculation stages as described in the function's description.

Reviewer 3: 'As described in description` sounds repetitive. 
 
```suggestion
1. Calculate the resulting value in the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function. Implement the necessary calculation stages. See the function's description for details.
```

> If your data set is large or you want to configure custom summaries on the server, refer to the [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation) article.

Reviewer: - "If your data set is large" - we haven't said that client-side summaries are for small datasets.
- "configure custom summaries on the server"- again, does "configure" means that summaries will be calculated on the server?

> The custom summary in this example calculates the sum of sale amounts for selected rows. To recalculate the resulting value on selection change, the code calls the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.

Reviewer 2: resulting value on selection change?
what are you trying to say?

Author: ![image](https://user-images.githubusercontent.com/66619750/110909032-a972d780-8320-11eb-8e1f-f735c8166503.png)
Each time a user selects/deselects an item, the summary ("the resulting value") is recalculated (please see the image).

> The custom summary in this example calculates the sum of sale amounts for selected rows. To recalculate the resulting value on selection change, the code calls the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.

Reviewer 2: ```suggestion
The custom summary in this example calculates the sum of sale amounts for selected rows. To recalculate the resulting value when selection is changed, the code calls the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.
```

> In this example, the custom summary displays the sum of sale amounts for selected rows only. To recalculate the resulting value on each selection change, the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method is called within the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.

Reviewer 3: I'm getting rid of the comma-separated clause and then the passive voice. 

```suggestion
The custom summary in this example calculates the sum of sale amounts for selected rows. To recalculate the resulting value on selection change, the code calls the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.
```

> Client-side custom summaries are suitable if your dataset is small. For large datasets, we recommend that you use [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation).

Reviewer 3: "Small" and "large" are vague terms. Can we maybe say something like this: 

```suggestion
Client-side custom summaries are suitable for small datasets. If your tests show that client-side calculations result in noticeable lags, we recommend that you use [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation).
```

DataGrid can calculate custom summaries on the client or server side. In this demo, we have implemented the client-side logic. The instructions below explain how to do this in your code:

> 1. If the **remoteOperations**.[summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/#summary), **remoteOperations**.[groupPaging](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/#groupPaging), or [remoteOperations](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/) property is used, set it to **false**.

Reviewer: ```suggestion
1. Make sure that the **remoteOperations**.[summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/#summary), **remoteOperations**.[groupPaging](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/#groupPaging), or [remoteOperations](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/remoteOperations/) property is not set or set to **false**.
```

1. Add a summary configuration object to the [summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/).[groupItems](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/groupItems/) or [summary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/).[totalItems](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/) array.

> 2. Set the summary object's [summaryType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#summaryType) to *"custom"*.

Reviewer: The word "summary" is repeated many times.

```suggestion
1. Set the object's [summaryType](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#summaryType) to *"custom"*.
```

3. Specify the custom summary's [name](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#name).         

> Use this value to identify the summary item when you calculate the resulting value.

Reviewer: ```suggestion
1. Specify the summary's [name](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#name). It will be used to identify the summary item within the **calculateCustomSummary** function in the next step.
```

1. Calculate the resulting value in the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function.          

> Implement the necessary calculation stages as described in the function's documentation.

Reviewer: ```suggestion
1. Calculate the resulting value in the [calculateCustomSummary](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/#calculateCustomSummary) function. Implement the necessary calculation stages as described in the function's description.
```

> In this example, the custom summary displays the sum of sale amounts for selected rows only. To rerender the resulting value on each selection change, the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler is implemented.

Reviewer: ```suggestion
In this example, the custom summary displays the sum of sale amounts for selected rows only. To recalculate the resulting value on each selection change, the [refresh()](/Documentation/ApiReference/UI_Components/dxDataGrid/Methods/#refresh) method is called within the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/#onSelectionChanged) handler.
```

> You should use client-side custom summaries if your data set is small. Otherwise, we recommend that you choose the server-side scenario. For details, refer to the [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation) article.

Reviewer: ```suggestion
Client-side custom summaries are suitable if your dataset is small. For large datasets, we recommend that you use [Server-Side Data Aggregation](/Documentation/Guide/UI_Components/DataGrid/Summaries/Custom_Aggregate_Function/#Server-Side_Data_Aggregation).
```
